File: EXAMPLES/FORTRAN/driver.f
C
C   PLAPACK General Linear Solver example application.
C
C   Purpose:   Implement the solution of a problem requiring
C              a general linear solver, including the problem
C              generation phase, the solution of the system,
C              and the retrieval of the solution data.
C
C   Datatypes: MPI_FLOAT, MPI_DOUBLE, MPI_DOUBLE_COMPLEX 
C               
C


      PROGRAM MAIN

      implicit none

      include "mpif.h"
      include "PLAf.h"
C  
C MPI_Comm
C     
      INTEGER COMM
C
C PLA_Template
C    
      INTEGER TEMPL
C
C PLA_Obj
C
      INTEGER 
     &     A,
     &     PIVOTS,
     &     X,
     &     B,
     &     B_NORM,
     &     INDEX,
     &     MINUS_ONE,
     &     ZERO,
     &     ONE
C
      DOUBLE PRECISION 
     &     OPERATION_COUNT, 
     &     B_NORM_VALUE, 
     &     TIME
      INTEGER 
     &     IERROR,
     &     SIZE, 
     &     NB_DISTR,
     &     ME,
     &     NPROCS,
     &     NPROWS,
     &     NPCOLS,
     &     DUMMY,
     &     INFO 
C
C MPI_Datatype 
C
      INTEGER DATATYPE


      COMM = MPI_COMM_NULL
      TEMPL = 0
      A = 0 
      PIVOTS = 0 
      X = 0 
      B = 0 
      B_NORM = 0 
      INDEX = 0
      MINUS_ONE = 0
      ZERO = 0
      ONE = 0

      CALL MPI_INIT( IERROR )

C
C Get problem size and distribution block size and broadcast 
C
      CALL MPI_COMM_RANK( MPI_COMM_WORLD, ME, IERROR )
      if ( ME .EQ. 0 ) THEN
         PRINT *, "-- enter processor mesh dimension ( rows cols ):"
         READ  *, NPROWS, NPCOLS
         PRINT *, "-- enter matrix size, distr. block size:"
         READ  *, SIZE, NB_DISTR
      ENDIF

      CALL MPI_BCAST( NPROWS,   1, MPI_INTEGER, 0, MPI_COMM_WORLD,
     &     IERROR )
      CALL MPI_BCAST( NPCOLS,   1, MPI_INTEGER, 0, MPI_COMM_WORLD,
     &     IERROR )
      CALL MPI_BCAST( SIZE,     1, MPI_INTEGER, 0, MPI_COMM_WORLD,
     &     IERROR )
      CALL MPI_BCAST( NB_DISTR, 1, MPI_INTEGER, 0, MPI_COMM_WORLD,
     &     IERROR )

C
C Two different ways to create a 2D communicator 
C
C  PLA_Comm1dTo2dRatio_F ( MPI_COMM_WORLD, 0.25d00, COMM )
      CALL PLA_Comm1dTo2d_F( MPI_COMM_WORLD, NPROWS, NPCOLS, 
     &     COMM, IERROR )
C
C Initialize PLAPACK 
C
      CALL PLA_INIT_F( COMM, IERROR )
C
C Create object distribution template 
C
      CALL PLA_TempCreate_F( NB_DISTR, 0, TEMPL, IERROR )
C
C Set the datatype : MPI_DOUBLE, MPI_FLOAT or MPI_DOUBLE_COMPLEX 
C
      DATATYPE = MPI_DOUBLE_PRECISION
C
C Create objects for problem to be solved 
C
      CALL PLA_MatrixCreate_F(  DATATYPE, SIZE, SIZE, TEMPL,
     &                           PLA_ALIGN_FIRST, PLA_ALIGN_FIRST, 
     &                           A, IERROR )

      CALL PLA_MvectorCreate_F( DATATYPE, SIZE, 1, 
     &                           TEMPL, PLA_ALIGN_FIRST, 
     &                           X, IERROR )

      CALL PLA_MvectorCreate_F( DATATYPE, SIZE, 1, 
     &                           TEMPL, PLA_ALIGN_FIRST, 
     &                           B, IERROR )

      CALL PLA_MvectorCreate_F( MPI_INTEGER, 
     &                           SIZE, 1, TEMPL, PLA_ALIGN_FIRST,
     &                           PIVOTS, IERROR )
C
C Create duplicated scalar constants with same datatype and template as A 
C
      CALL PLA_CreateConstantsConfTo_F( A, MINUS_ONE, ZERO, ONE, 
     &     IERROR )
C
C Create a problem to be solved: A x = b 
C
      CALL CREATE_PROBLEM_F( A, X, B )
C
C Start timing 
      CALL MPI_BARRIER( MPI_COMM_WORLD, IERROR )
      TIME = MPI_WTIME( )
C
C Factor P A -> L U overwriting lower triangular portion of A with L, upper, U 
C
      CALL PLA_LU_F( A, PIVOTS, INFO )

      IF ( INFO .NE. 0 ) THEN
         PRINT *, "Zero pivot encountered at row", INFO
      ELSE
C
C  Apply the permutations to the right hand sides 
C
         CALL PLA_ApplyPivotsToRows_F ( B, PIVOTS, IERROR )
C
C Solve L y = b, overwriting b with y 
C
         CALL PLA_Trsv_F( PLA_LOWER_TRIANGULAR, PLA_NO_TRANSPOSE, 
     &                    PLA_UNIT_DIAG, A, B, IERROR )
C
C Solve U x = y (=b), overwriting b with x
C
         CALL PLA_Trsv_F( PLA_UPPER_TRIANGULAR, PLA_NO_TRANSPOSE,
     &                    PLA_NONUNIT_DIAG, A, B, IERROR )
C
C Stop timing 
C
         CALL MPI_BARRIER( MPI_COMM_WORLD, IERROR )
         TIME = MPI_WTIME() - TIME
C
C Report performance
C
         IF ( ME .EQ. 0 ) THEN
            CALL MPI_COMM_SIZE(MPI_COMM_WORLD, NPROCS, IERROR )

            OPERATION_COUNT = 2.0/3.0 * SIZE * SIZE * SIZE

            IF ( MPI_DOUBLE_COMPLEX .EQ. DATATYPE ) 
     &           OPERATION_COUNT = 4 * OPERATION_COUNT

            PRINT *, "N = ", SIZE, " TIME = ", TIME, " MFLOPS/node = ", 
     &           OPERATION_COUNT / TIME * 1.0e-6 / NPROCS
         ENDIF
C
C    /* Process the answer.  As an example, this routine brings 
C       result x (stored in b) to processor 0 and prints first and 
C       last entry */
C
         CALL PROCESS_ANSWER_F( B )
C
C    /* Check answer by overwriting b <- b - x (where b holds computed
C       approximation to x) */
C     
         CALL PLA_Axpy_F( MINUS_ONE, X, B, IERROR )
C
C    /* Create 1x1 multiscalars to hold largest (in abs. value) element 
C       of b - x and index of largest value */
C
         CALL PLA_MscalarCreate_F( DATATYPE, 
     &                              PLA_ALL_ROWS, PLA_ALL_COLS,
     &                              1, 1, TEMPL, B_NORM, IERROR )
C
         CALL PLA_MscalarCreate_F( MPI_INTEGER, 
     &                              PLA_ALL_ROWS, PLA_ALL_COLS,
     &                              1, 1, TEMPL, INDEX, IERROR )
C
         CALL PLA_Iamax_F ( B, INDEX, B_NORM, IERROR )
C
C    /* Report norm of b - x */
C
         IF ( ME .EQ. 0 ) THEN
            CALL PLA_ObjGetLocalContents_F( 
     &           B_NORM, PLA_NO_TRANS, DUMMY, DUMMY,
     &           B_NORM_VALUE, 1, 1, IERROR )
            PRINT *, "Largest element in (computed x) - x : ", 
     &               B_NORM_VALUE
         ENDIF
      ENDIF
C
C  /* Free the linear algebra objects */
C
      CALL PLA_ObjFree_F( A, IERROR )
      CALL PLA_ObjFree_F( X, IERROR )
      CALL PLA_ObjFree_F( B, IERROR )
      CALL PLA_ObjFree_F( B_NORM, IERROR )
      CALL PLA_ObjFree_F( PIVOTS, IERROR )
      CALL PLA_ObjFree_F( INDEX, IERROR )
      CALL PLA_ObjFree_F( MINUS_ONE, IERROR )
      CALL PLA_ObjFree_F( ZERO, IERROR )
      CALL PLA_ObjFree_F( ONE, IERROR )
C
C  /* Free the template */
C
      CALL PLA_TempFree_F( TEMPL, IERROR )
C
C  /* Finalize PLAPACK and MPI */
C
      CALL PLA_Finalize_F(IERROR )
      CALL MPI_Finalize( IERROR )

      STOP
      END

File: EXAMPLES/LU/driver.c
/*
   PLAPACK Release 3.0
   
   Copyright (C) 2000
   Robert A. van de Geijn and The University of Texas at Austin

   For GNU-license details see the file GNU_license in the 
   PLAPACK root directory.
*/

/*
 *   PLAPACK General Linear Solver example application.
 *
 *   Purpose:   Implement the solution of a problem requiring
 *              a general linear solver, including the problem
 *              generation phase, the solution of the system,
 *              and the retrieval of the solution data.
 *
 *   Datatypes: MPI_FLOAT, MPI_DOUBLE, MPI_DOUBLE_COMPLEX 
 *               
 */

#include "PLA.h"                 /* PLAPACK header file */   

#if MANUFACTURE == CRAY
#include <mpp/rastream.h>
#endif  

int main(int argc, char *argv[])
{
  /* Declarations */
  MPI_Comm 
    comm;

  PLA_Template 
    templ = NULL;

  PLA_Obj  
    A = NULL,           
    pivots = NULL,
    x = NULL,         
    b = NULL, 
    b_norm = NULL,
    index = NULL,
    minus_one = NULL;

  double 
    operation_count,
    b_norm_value, 
    time;

  int  
    size, 
    nb_distr, nb_alg, 
    me, nprocs, 
    nprows, npcols,
    dummy, 
    ierror,
    info = 0;
  
  MPI_Datatype 
    datatype;

  /* Initialize MPI */
  MPI_Init(&argc, &argv);

#if MANUFACTURE == CRAY
  set_d_stream( 1 );
#endif

  /* Get problem size and distribution block size and broadcast */
  MPI_Comm_rank(MPI_COMM_WORLD, &me);
  if (0 == me) {
    printf("enter processor mesh dimension ( rows cols ):\n");
    scanf("%d %d", &nprows, &npcols );
    printf("enter matrix size, distr. block size:\n");
    scanf("%d %d", &size, &nb_distr );
    printf("enter algorithmic blocksize:\n");
    scanf("%d", &nb_alg );
    printf("Turn on error checking? (1 = YES, 0 = NO):\n");
    scanf("%d", &ierror );
  }

  MPI_Bcast(&nprows,   1, MPI_INT, 0, MPI_COMM_WORLD);
  MPI_Bcast(&npcols,   1, MPI_INT, 0, MPI_COMM_WORLD);
  MPI_Bcast(&size,     1, MPI_INT, 0, MPI_COMM_WORLD);
  MPI_Bcast(&nb_distr, 1, MPI_INT, 0, MPI_COMM_WORLD);
  MPI_Bcast(&nb_alg, 1, MPI_INT, 0, MPI_COMM_WORLD);
  MPI_Bcast(&ierror, 1, MPI_INT, 0, MPI_COMM_WORLD);

  if ( ierror ) 
    PLA_Set_error_checking( ierror, TRUE, TRUE, FALSE );
  else
    PLA_Set_error_checking( ierror, FALSE, FALSE, FALSE );

  pla_Environ_set_nb_alg (PLA_OP_ALL_ALG,
			  nb_alg);

  /* Two different ways to create a 2D communicator */
/*  PLA_Comm_1D_to_2D_ratio(MPI_COMM_WORLD, 0.25, &comm); */ 
  PLA_Comm_1D_to_2D(MPI_COMM_WORLD, nprows, npcols, &comm); 

  /* Initialize PLAPACK */
  PLA_Init(comm);

  /* Create object distribution template */
  PLA_Temp_create( nb_distr, 0, &templ );

  /* Set the datatype : MPI_DOUBLE, MPI_FLOAT or MPI_DOUBLE_COMPLEX */
  datatype = MPI_DOUBLE_COMPLEX;

  /* Create objects for problem to be solved */
  PLA_Matrix_create(  datatype, size, size, 
                      templ, PLA_ALIGN_FIRST, PLA_ALIGN_FIRST, &A );

  PLA_Mvector_create( datatype, size, 1, templ, PLA_ALIGN_FIRST, &x );

  PLA_Mvector_create( datatype, size, 1, templ, PLA_ALIGN_FIRST, &b );
   
  PLA_Mvector_create( MPI_INT, size, 1, templ, PLA_ALIGN_FIRST, &pivots );
   
  /* Create 1x1 multiscalars to hold largest (in abs. value) element 
     of b - x and index of largest value */
  if ( datatype == MPI_DOUBLE_COMPLEX || datatype == MPI_DOUBLE )
    PLA_Mscalar_create( MPI_DOUBLE,
			 PLA_ALL_ROWS, PLA_ALL_COLS,
			 1, 1, templ, &b_norm );
  else   if ( datatype == MPI_COMPLEX || datatype == MPI_FLOAT )
    PLA_Mscalar_create( MPI_FLOAT,
			 PLA_ALL_ROWS, PLA_ALL_COLS,
			 1, 1, templ, &b_norm );

  /* Create duplicated scalar constants with same datatype and template as A */
  PLA_Create_constants_conf_to( A, &minus_one, NULL, NULL );

  /* Create a problem to be solved: A x = b */
  create_problem( A, x, b );

  /* Start timing */
  MPI_Barrier( MPI_COMM_WORLD );
  time = MPI_Wtime( );

  /* Factor P A -> L U overwriting lower triangular portion of A with L, upper, U */
  info = PLA_LU ( A, pivots);

  if ( info != 0 ) {
    printf("Zero pivot encountered at row %d.\n", info);
  }
  else {
    /* Apply the permutations to the right hand sides */
    PLA_Apply_pivots_to_rows ( b, pivots);

    /* Solve L y = b, overwriting b with y */
    PLA_Trsv( PLA_LOWER_TRIANGULAR, PLA_NO_TRANSPOSE, PLA_UNIT_DIAG, A, b );

    /* Solve U x = y (=b), overwriting b with x */
    PLA_Trsv( PLA_UPPER_TRIANGULAR, PLA_NO_TRANSPOSE,  PLA_NONUNIT_DIAG, A, b );

    /* Stop timing */
    MPI_Barrier( MPI_COMM_WORLD );
    time = MPI_Wtime() - time;

    /* Report performance */
    if ( me == 0 ) {
      MPI_Comm_size(MPI_COMM_WORLD, &nprocs);

      operation_count = 2.0/3.0 * size * size * size;

      if ( MPI_DOUBLE_COMPLEX == datatype )
	operation_count *= 4;

      switch( datatype ){
      case MPI_DOUBLE_COMPLEX: 
	printf( "datatype timed: double precision complex\n"); 
	printf( "(op count = 8/3 n^3)\n"); 
	break;
      case MPI_DOUBLE: 
	printf( "datatype timed: double precision\n"); 
	printf( "(op count = 2/3 n^3)\n"); 
	break;
      case MPI_COMPLEX: 
	printf( "datatype timed: single precision complex\n"); 
	printf( "(op count = 8/3 n^3)\n"); 
	break;
      case MPI_FLOAT: 
	printf( "datatype timed: single precision\n"); 
	printf( "(op count = 2/3 n^3)\n"); 
	break;
      }

      printf("n = %d, time = %lf, MFLOPS/node = %lf\n", size, time,
	      operation_count / time * 1.0e-6 / nprocs );
    }

    /* Process the answer.  As an example, this routine brings 
       result x (stored in b) to processor 0 and prints first and 
       last entry */
    Process_answer( b ); 

    /* Check answer by overwriting b <- b - x (where b holds computed
       approximation to x) */

    PLA_Axpy( minus_one, x, b );

    PLA_Nrm2( b, b_norm);

    /* Report norm of b - x */
    if ( me == 0 ) {
      PLA_Obj_get_local_contents( b_norm, PLA_NO_TRANS, &dummy, &dummy,
				  &b_norm_value, 1, 1 );
      printf( "Norm2 of x - computed x : %le\n", b_norm_value );
    }
  } 

  /* Free the linear algebra objects */
  PLA_Obj_free(&A);            PLA_Obj_free(&x);
  PLA_Obj_free(&b);            PLA_Obj_free(&minus_one);
  PLA_Obj_free(&b_norm);       PLA_Obj_free (&pivots);

  /* Free the template */
  PLA_Temp_free(&templ);

  /* Finalize PLAPACK and MPI */
  PLA_Finalize( );
  MPI_Finalize( );
}




